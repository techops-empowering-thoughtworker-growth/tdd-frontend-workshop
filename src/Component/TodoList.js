import React from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';

const TodoList = (props) => {
  const { todoItems, isCompleted, setTodoItems } = props;
  const filteredTodoItems = isCompleted
    ? todoItems.filter((task) => task.completed)
    : todoItems.filter((task) => !task.completed);

  const handleDelete = (e) => {
    const deleteId = e.target.id;
    axios.delete(`http://localhost:8080/tasks/${deleteId}`).then(() => {
      setTodoItems(todoItems.filter((item) => item.id !== +deleteId));
    });
  };

  const handleUpdate = (updateId) => {
    const taskToUpdate = todoItems.find((item) => item.id === +updateId);
    axios
      .put(`http://localhost:8080/tasks/${updateId}`, {
        name: taskToUpdate.name,
        completed: !taskToUpdate.completed,
      })
      .then((res) =>
        setTodoItems(
          todoItems.map((item) => (item.id === +updateId ? res.data : item))
        )
      );
  };

  return (
    <ul className={isCompleted ? 'completed-list' : 'todo-list'}>
      {isCompleted === true && <h2>Completed</h2>}
      {filteredTodoItems.map((todo) => (
        <li className={'task-item'} key={todo.id}>
          <input
            type={'checkbox'}
            defaultChecked={isCompleted}
            id={todo.id}
            onClick={() => handleUpdate(todo.id)}
          />
          <p className={'task-text'}>{todo.name}</p>
          <i className={'bi-trash'} id={todo.id} onClick={handleDelete}></i>
        </li>
      ))}
    </ul>
  );
};

TodoList.propTypes = {
  todoItems: PropTypes.array.isRequired,
  handleChangeProps: PropTypes.func,
  deleteTodoProps: PropTypes.func,
  setUpdate: PropTypes.func,
};

export default TodoList;
