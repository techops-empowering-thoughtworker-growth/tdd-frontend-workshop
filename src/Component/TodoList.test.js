import { render, mount } from 'enzyme';
import axios from 'axios';
import { act } from 'react-dom/test-utils';
import TodoList from './TodoList';

const props = {
  todoItems: [
    { id: 1, name: 'task 01', completed: true },
    { id: 2, name: 'task 02', completed: false },
  ],
  setTodoItems: jest.fn(),
  isCompleted: true,
};

jest.mock('axios');

describe('TodoList', () => {
  it('should take TodoList snapshot', () => {
    const wrapper = render(<TodoList {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should list correctly using Enzyme and Jest', () => {
    let wrapper;
    wrapper = mount(<TodoList {...props} />);
    wrapper.update();
    expect(wrapper.find('.task-item')).toHaveLength(1);
    expect(wrapper.find('.task-item').at(0).find('.task-text').text()).toBe(
      'task 01'
    );
  });

  it('should delete the item on delete button click', async () => {
    let wrapper;
    axios.delete.mockResolvedValue();

    await act(async () => {
      wrapper = mount(<TodoList {...props} />);
    });
    wrapper.update();

    await act(async () => {
      wrapper
        .find('.task-item')
        .at(0)
        .simulate('mouseover')
        .find('.bi-trash')
        .simulate('click');
    });
    wrapper.update();

    expect(axios.delete).toHaveBeenCalledWith('http://localhost:8080/tasks/1');
  });
});
