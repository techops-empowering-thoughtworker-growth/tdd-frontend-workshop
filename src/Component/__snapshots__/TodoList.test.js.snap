// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should take TodoList snapshot 1`] = `
LoadedCheerio {
  "0": <ul
    class="completed-list"
  >
    <h2>
      Completed
    </h2>
    <li
      class="task-item"
    >
      <input
        checked=""
        id="1"
        type="checkbox"
      />
      <p
        class="task-text"
      >
        task 01
      </p>
      <i
        class="bi-trash"
        id="1"
      />
    </li>
  </ul>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
